### ISSUE TITLE
[2.c.ii] For these instructions, list which stages need to be stalled and which stages need to be squashed/flushed relative to the stage each of these instructions is in.

### ISSUE THREAD

**Issue by aidanfoss** [conneroisu/sfhw-proj2#18]

**Comment by aidanfoss** [comment]
related to #23

**Comment by conneroisu** [comment]
For the instructions BNE, BEQ, J, JAL, and JR that result in a non-sequential PC update, the following stages need to be stalled and squashed/flushed relative to the stage each instruction is in:

Stalling:

- Stalling occurs for 1 cycle.
- The IF/ID pipeline register is stalled to prevent the next instruction from entering the Decode (ID) stage.
- The ID/EX pipeline register is stalled to prevent the instruction in the Decode (ID) stage from moving to the Execute (EX) stage.

Flushing:

- Flushing occurs for 1 cycle.
- Flushing takes place in the Fetch (IF) stage to discard the instruction that was fetched based on the sequential PC.

**Comment by conneroisu** [comment]
1. Instructions producing values on the ALU Result signal in the execute stage:

   - add, addi, addiu, addu
   - sub, subu
   - and, andi
   - or, ori
   - xor, xori
   - nor
   - slt, slti
   - sll, srl, sra
   - sllv, srlv, srav

2. Instructions producing/using values on the mem data signal in the memory stage:

   - lw, lh, lhu, lb, lbu (load instructions)
   - sw, sh, sb (store instructions)

3. Instructions producing values on the branchLogic signal going into the fetch logic unit:

   - beq, bne

4. Instructions producing values on the JumpLogic and JregLogic signals going into the fetch logic unit:
   - j, jal
   - jr (uses iJreg signal specifically for register jumping)

**Comment by conneroisu** [comment]
For the instructions BNE, BEQ, J, JAL, and JR that result in a non-sequential PC update, the following stages need to be stalled and squashed/flushed relative to the stage each instruction is in:

Stalling:

- Stalling occurs for 1 cycle.
- The IF/ID pipeline register is stalled to prevent the next instruction from entering the Decode (ID) stage.
- The ID/EX pipeline register is stalled to prevent the instruction in the Decode (ID) stage from moving to the Execute (EX) stage.

Flushing:

- Flushing occurs for 1 cycle.
- Flushing takes place in the Fetch (IF) stage to discard the instruction that was fetched based on the sequential PC.

In summary, when any of these instructions (BNE, BEQ, J, JAL, JR) are encountered, the pipeline is stalled for 1 cycle, and the Fetch (IF) stage is flushed simultaneously to ensure that the next instruction executed is based on the updated non-sequential PC.

**Comment by aidanfoss** [comment]
The answer for this one was half-assed. Perhaps would be good to double check. Right before I commented this, I pasted conner's previous comments into the report.
Likely has the information the question is asking for, good to double check and make more concise

**Comment by aidanfoss** [comment]
![image](https://github.com/user-attachments/assets/1097f886-9fe5-4e37-bf2a-fd4ef4a5031c)

**Comment by aidanfoss** [comment]
https://docs.google.com/spreadsheets/d/1RUNOOiJ3DZ2heZ5_la7ENw52pGrtweVW7ax-dkXzsRU/edit?gid=504132693#gid=504132693
